src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -2119,6 +2119,38 @@ void MergeTreeData::dropAllData()
 
     LOG_TRACE(log, "dropAllData: removing data from filesystem.");
 
+    for (const auto & disk : getDisks())
+    {
+        if (disk->isBroken())
+            continue;
+
+        try
+        {
+            if (disk->supportZeroCopyReplication())
+            {
+                auto detached_path = fs::path(relative_data_path) / MergeTreeData::DETACHED_DIR_NAME;
+                for (auto it = disk->iterateDirectory(detached_path); it->isValid(); it->next())
+                {
+                    if (disk->isDirectory(detached_path / it->name()))
+                    {
+                        // Avoid removing shared remote objects when table dropped only on one replica
+                        removeDetachedPart(disk, detached_path / it->name() / "", it->name(), false);
```

Comment:
It avoids only a subset of possible issues. Imagine the following situation: 
1. Table is enqueued for cleanup after drop, data is not deleted from disk yet
2. Server is restarted
3. DatabaseCatalog is trying to load dropped table in order to call `drop()` and cleanup everything gracefully
4. It fails to load the table for some reason
5. It falls back to recursive removal and simply calls `disk->removeRecursive(table_data_path)`
6. Seems like DIskObjectStorage will remove shared data